File: EvaluateReversePolishNotation.py
# EvaluateReversePolishNotation


class Solution:
    # @param tokens, a list of string
    # @return an integer
    def evalRPN(self, tokens):
        storage=[]
        size = len(tokens)
        for i in range(0,size):
            if tokens[i] in ['+','*','/','-']:
                a2 = storage[-1]
                storage.pop(-1)
                a1 = storage[-1]
                storage.pop(-1)
                if tokens[i][0] == '+':                   
                    result = a1+a2
                    storage.append(result)
                elif tokens[i][0] == '-':
                    result = a1-a2
                    storage.append(result)
                elif tokens[i][0] == '*':
                    result = a1*a2
                    storage.append(result)
                else:
                    if a1%a2 != 0 and a1/a2< 0:
                        result = a1/a2+1
                    else:
                        result = a1/a2
                    storage.append(int(result))             
            else:
                storage.append(int(tokens[i]))
        else:
            result = storage[0]

        return result
        

if __name__ == "__main__":
    s=Solution()
    tokens=["3","-4","+"]
    t=s.evalRPN(tokens)
    print(t)

File: EvaluateReversePolishNotation.cpp
#include <iostream>
#include <string>
#include <vector>
#include <sstream>

using namespace std;

class Solution {
public:
    int evalRPN(vector<string> &tokens) {
      
		int result,temp;
		int vector_size = tokens.size();
		int i;
		vector<int> storage;
		for(i=0;i<vector_size;i++){
		
			
			if((tokens.at(i)[0]!='/')&&(tokens.at(i)[0]!='*')&&(tokens.at(i)[0]!='+')&&(tokens.at(i)[0]!='-'||(tokens.at(i)[0]=='-'&&tokens.at(i).length()!=1))){		
				stringstream ss;
				ss<<tokens.at(i);
				ss>>temp;
				storage.push_back(temp);			
			}else{
			    int a1,a2;
				switch(tokens.at(i)[0]){			
				case '+':a2=storage.back();storage.pop_back();a1=storage.back();storage.pop_back(); result = a1+a2;storage.push_back(result);break;
				case '-':a2=storage.back();storage.pop_back();a1=storage.back();storage.pop_back(); result = a1-a2;storage.push_back(result);break;
				case '*':a2=storage.back();storage.pop_back();a1=storage.back();storage.pop_back(); result = a1*a2;storage.push_back(result);break;
				case '/':a2=storage.back();storage.pop_back();a1=storage.back();storage.pop_back(); result = a1/a2;storage.push_back(result);break;
				}
			}
		}

		result = storage.back();
		return result;
    }
};


int main(){

	//"2", "1", "+", "3", "*"
	vector<string> tokens(1);
	tokens.at(0)="3";
	//tokens.at(1)="-4";
	//tokens.at(2)="+";
/*	tokens.at(3)="3";
	tokens.at(4)="/"*/;

	Solution so;
	int result=so.evalRPN(tokens); 
	cout<<result<<endl;
	
	return 0;

}
